src/plugins/intel_cpu/src/memory_state.cpp

```diff
@@ -146,50 +154,123 @@ MemoryPtr VariableStateDoubleBuffer::internal_state_mem() const {
     return prime_mem();
 }
 
-VariableStateSingleBuffer::VariableStateSingleBuffer(const std::string& name,
-                                                     const MemoryPtr& buffer,
-                                                     const MemoryDescPtr& external_desc) :
-    VariableStateBase(name, external_desc) {
-    OPENVINO_ASSERT(buffer);
-    m_internal_mem = buffer;
-    m_internal_desc = m_internal_mem->getDescPtr();
-    auto&& shape = m_internal_desc->getShape();
-    //TODO what if by some reason we already have internal static state while the node is dynamic, is it even possible?
+VariableStateKVcache::VariableStateKVcache(
+    const std::string& name,
+    const MemoryDescPtr& external_desc,
+    const BlockedMemoryDescPtr& dense_internal_desc) :
+    VariableStateBase(name, external_desc), m_dense_internal_desc(dense_internal_desc) {
+    auto&& shape = external_desc->getShape();
 
-    if (shape.isStatic()) {
-        m_internal_mem->nullify();
-    } else {
-        //in the case of the original desc has dynamic shape we create an empty tensor
-        auto new_desc = to_static(m_internal_desc);
-        m_internal_mem->redefineDesc(new_desc);
+    OPENVINO_ASSERT(shape.isDynamic(), "VariableStateKVcache is unexpectedly initalized with a static tensor");
+}
+
+ov::SoPtr<ov::ITensor> VariableStateKVcache::get_state() const {
+    auto actual_internal_desc = m_internal_mem->getDescWithType<BlockedMemoryDesc>();
+    auto&& dims = actual_internal_desc->getShape().getStaticDims();
+
+    auto actual_external_desc = get_external_desc()->cloneWithNewDims(dims);
+    auto external_mem = std::make_shared<Memory>(get_engine(), actual_external_desc);
+
+    // let's assume 4th rank KV tensors. This may be extended later
+    OPENVINO_ASSERT(actual_internal_desc->getShape().getRank() == 4);
+    OPENVINO_ASSERT(actual_external_desc->getShape().getRank() == 4);
+
+    auto&& actual_internal_order = actual_internal_desc->getOrder();
+    //sanity check
+    OPENVINO_ASSERT(actual_internal_order == m_dense_internal_desc->getOrder());
+
+    // Warning, this implementation is very KV cache specific it assumes that S is always a last dimension and it's not
+
+    if (m_hidden_state) {
```

Comment:
>>> what if m_hidden_state is empty?
It should be an invalid state if it's empty and get_state is called. OPENVINO_ASSERT is added.

>>> when m_hidden_state will be empty?
Only no infer and set_state is called, it will be empty.

>>> what will get_state() return after reset?
It should be an undefined behavior. Adding OPENVINO_ASSERT to notify the user.